cvxpy/tests/solver_test_helpers.py

```diff
@@ -110,27 +110,120 @@ def check_complementarity(self, places) -> None:
                                   cp.constraints.SOC,
-                                  cp.constraints.NonPos,
-                                  cp.constraints.Zero)):
+                                  cp.constraints.NonNeg,
+                                  cp.constraints.Zero,
+                                  cp.constraints.PSD,
+                                  cp.constraints.PowCone3D)):
                 comp = cp.scalar_product(con.args, con.dual_value).value
-            elif isinstance(con, cp.constraints.PowCone3D):
-                comp = cp.scalar_product(con.args[:3], con.dual_value).value
-            elif isinstance(con, (cp.constraints.Inequality,
-                                  cp.constraints.Equality)):
-                comp = cp.scalar_product(con.expr, con.dual_value).value
             elif isinstance(con, cp.constraints.PowConeND):
                 msg = '\nPowConeND dual variables not implemented;' \
                        + '\nSkipping complementarity check.'
                 warnings.warn(msg)
             else:
                 raise ValueError('Unknown constraint type %s.' % type(con))
             self.tester.assertAlmostEqual(comp, 0, places)
+            
+    def check_stationary_lagrangian(self, places) -> None:
+        L = self.prob.objective.expr
+        for con in self.constraints:
+            if isinstance(con, (cp.constraints.Inequality,
+                                cp.constraints.Equality)):
+                dual_var_value = con.dual_value
+                prim_var_expr = con.expr
+                if self.prob.objective.NAME == 'minimize':
+                    L = L + cp.scalar_product(dual_var_value, prim_var_expr)
+                else:
+                    L = L - cp.scalar_product(dual_var_value, prim_var_expr)
+            elif isinstance(con, (cp.constraints.ExpCone,
+                                  cp.constraints.SOC,
+                                  cp.constraints.Zero,
+                                  cp.constraints.NonNeg,
+                                  cp.constraints.PSD,
+                                  cp.constraints.PowCone3D)):
+                if self.prob.objective.NAME == 'minimize':
+                    L = L - cp.scalar_product(con.args, con.dual_value)
+                else:
+                    L = L + cp.scalar_product(con.args, con.dual_value)
+            else:
+                raise NotImplementedError()
+        g = L.grad
+        bad_norms = []
+
+        """The reason 'bad' norms could arise (outside of an error by CVXPY) would be if a
+        constraint was introduced in a variable via a flag during it's declaration and not
+        explicitly in the list of constraints passed to the problem --- thus making the above
+        constructed lagrangian incorrect. In these cases instead of checking if the dLdX = 0
+        we check if dLdX \in K^{*} (i.e. the dual cone of the implicit constraint)"""
```

Comment:
This comment isn't quite right. The Lagrangian is defined by _conventions_. Once we've set certain conventions, it's up to us to interpret the Lagrangian correctly.  Our specific conventions are that conic constraints declared with Variable attributes (PSD=True, diag=True, etc...) are not moved into the Lagrangian. The effect of this convention is that we don't check if the gradient of the Lagrangian is equal to zero, we check if ``dLdX \in K^{*}``. This captures the case when ``K = R^n`` since the dual to ``R^n`` is the singleton set that only contains the zero vector.